// decentraland-ecs/artifacts/amd.js
// A naive attempt at getting the global `this`. Don’t use `this`!
var getGlobalThis = function () {
    // @ts-ignore
    if (typeof globalThis !== 'undefined')
        return globalThis;
    // @ts-ignore
    if (typeof self !== 'undefined')
        return self;
    // @ts-ignore
    if (typeof window !== 'undefined')
        return window;
    // Note: this might still return the wrong result!
    if (typeof this !== 'undefined')
        return this;
    throw new Error('Unable to locate global `this`');
};
var globalObject = getGlobalThis();
var loader;
(function (loader) {
    'use strict';
    var MODULE_LOADING = 1;
    var MODULE_READY = 2;
    var anonymousQueue = [];
    var settings = {
        baseUrl: ''
    };
    var registeredModules = {};
    function config(config) {
        if (typeof config === 'object') {
            for (var x in config) {
                if (config.hasOwnProperty(x)) {
                    settings[x] = config[x];
                }
            }
        }
    }
    loader.config = config;
    function createModule(name, context, handlers) {
        if (handlers === void 0) { handlers = []; }
        return {
            name: name,
            dclamd: MODULE_LOADING,
            handlers: handlers,
            context: context,
            exports: {}
        };
    }
    function define(id, dependencies, factory) {
        var argCount = arguments.length;
        if (argCount === 1) {
            factory = id;
            dependencies = ['require', 'exports', 'module'];
            id = null;
        }
        else if (argCount === 2) {
            if (settings.toString.call(id) === '[object Array]') {
                factory = dependencies;
                dependencies = id;
                id = null;
            }
            else {
                factory = dependencies;
                dependencies = ['require', 'exports', 'module'];
            }
        }
        if (!id) {
            anonymousQueue.push([dependencies, factory]);
            return;
        }
        function ready() {
            var handlers, context;
            if (registeredModules[id]) {
                handlers = registeredModules[id].handlers;
                context = registeredModules[id].context;
            }
            else {
                registeredModules[id] = createModule(id);
            }
            var module = registeredModules[id];
            module.exports =
                typeof factory === 'function'
                    ? factory.apply(null, anonymousQueue.slice.call(arguments, 0)) || registeredModules[id].exports || {}
                    : factory;
            module.dclamd = MODULE_READY;
            module.context = context;
            for (var x = 0, xl = handlers ? handlers.length : 0; x < xl; x++) {
                handlers[x](module.exports);
            }
        }
        if (!registeredModules[id])
            registeredModules[id] = createModule(id);
        updateExistingModuleIfIndexModule(id);
        registeredModules[id].dependencies = dependencies.map(function (dep) { return _toUrl(dep, id); });
        require(dependencies, ready, id);
    }
    loader.define = define;
    function updateExistingModuleIfIndexModule(id) {
        var _a;
        var idParts = id.split('/');
        if (idParts[idParts.length - 1] === 'index' && idParts.length > 1) {
            idParts.pop();
            var noIndexId = idParts.join('/');
            var existingModule = registeredModules[noIndexId];
            registeredModules[noIndexId] = registeredModules[id];
            if (existingModule) {
                registeredModules[noIndexId].context = existingModule.context;
                (_a = registeredModules[noIndexId].handlers).push.apply(_a, existingModule.handlers);
            }
        }
    }
    (function (define) {
        define.amd = {};
    })(define = loader.define || (loader.define = {}));
    function hasDependencyWith(moduleId, otherModuleId) {
        if (!registeredModules[moduleId] || !registeredModules[moduleId].dependencies) {
            return false;
        }
        var directDependencies = registeredModules[moduleId].dependencies;
        // Dependencies can be viewed as a graph. We keep track of the transitions between nodes that we have visited (to avoid infinite loop),
        // and those that we want to visit next. We want to find all the nodes reachable from moduleId, to see if those inclued otherModuleId
        var visited = [moduleId, 'require', 'exports', 'module'];
        // If the dependency is one of those automatically assumed visited, we know we have a dependency
        if (visited.indexOf(otherModuleId) !== -1)
            return true;
        var toVisit = directDependencies.filter(function (dep) { return visited.indexOf(dep) === -1; });
        while (toVisit.length > 0) {
            // If among the nodes we want to visit we have the module, then we know we have a dependency
            if (toVisit.indexOf(otherModuleId) !== -1)
                return true;
            var dependencyId = toVisit.shift();
            visited.push(dependencyId);
            var module = registeredModules[dependencyId];
            if (module && module.dependencies) {
                for (var i = 0; i < module.dependencies.length; i++) {
                    var moduleDependencyId = module.dependencies[i];
                    if (visited.indexOf(moduleDependencyId) === -1 && toVisit.indexOf(moduleDependencyId) === -1) {
                        toVisit.push(moduleDependencyId);
                    }
                }
            }
        }
        // If we have visited all the graph and we didn't find a dependency, then as far as we know, we don't have a dependency yet
        return false;
    }
    function require(modules, callback, context) {
        var loadedModulesExports = [];
        var loadedCount = 0;
        var hasLoaded = false;
        if (typeof modules === 'string') {
            if (registeredModules[modules] && registeredModules[modules].dclamd === MODULE_READY) {
                return registeredModules[modules];
            }
            throw new Error(modules + ' has not been defined. Please include it as a dependency in ' + context + "'s define()");
        }
        var xl = modules.length;
        var _loop_1 = function (x) {
            switch (modules[x]) {
                case 'require':
                    var _require = function (new_module, callback) {
                        return require(new_module, callback, context);
                    };
                    _require.toUrl = function (module) {
                        return _toUrl(module, context);
                    };
                    loadedModulesExports[x] = _require;
                    loadedCount++;
                    break;
                case 'exports':
                    loadedModulesExports[x] = registeredModules[context].exports;
                    loadedCount++;
                    break;
                case 'module':
                    loadedModulesExports[x] = {
                        id: context,
                        uri: _toUrl(context)
                    };
                    loadedCount++;
                    break;
                default:
                    // If we have a circular dependency, then we resolve the module even if it hasn't loaded yet
                    if (hasDependencyWith(modules[x], context)) {
                        loadedModulesExports[x] = registeredModules[modules[x]].exports;
                        loadedCount++;
                    }
                    else {
                        load(modules[x], function (loadedModuleExports) {
                            loadedModulesExports[x] = loadedModuleExports;
                            loadedCount++;
                            if (loadedCount === xl && callback) {
                                hasLoaded = true;
                                callback.apply(null, loadedModulesExports);
                            }
                        }, context);
                    }
            }
        };
        for (var x = 0; x < xl; x++) {
            _loop_1(x);
        }
        if (!hasLoaded && loadedCount === xl && callback) {
            callback.apply(null, loadedModulesExports);
        }
    }
    loader.require = require;
    function createMethodHandler(rpcHandle, method) {
        return function () {
            return dcl.callRpc(rpcHandle, method.name, anonymousQueue.slice.call(arguments, 0));
        };
    }
    function load(moduleName, handler, context) {
        moduleName = context ? _toUrl(moduleName, context) : moduleName;
        if (registeredModules[moduleName]) {
            if (registeredModules[moduleName].dclamd === MODULE_LOADING) {
                handler && registeredModules[moduleName].handlers.push(handler);
            }
            else {
                handler && handler(registeredModules[moduleName].exports);
            }
        }
        else {
            var module_1 = (registeredModules[moduleName] = createModule(moduleName, context, [handler]));
            if (moduleName.indexOf('@') === 0) {
                if (typeof dcl !== 'undefined') {
                    dcl.loadModule(moduleName).then(function (descriptor) {
                        var createdModuleExports = {};
                        for (var i in descriptor.methods) {
                            var method = descriptor.methods[i];
                            createdModuleExports[method.name] = createMethodHandler(descriptor.rpcHandle, method);
                        }
                        // This is somewhat repeated with the ready function above... Should refactor with clear head
                        module_1.dclamd = MODULE_READY;
                        module_1.exports = createdModuleExports;
                        var handlers = module_1.handlers;
                        for (var i = 0; i < handlers.length; i++) {
                            handlers[i](createdModuleExports);
                        }
                    });
                }
            }
        }
    }
    if (typeof dcl !== 'undefined') {
        dcl.onStart(function () {
            var notLoadedModules = [];
            for (var i in registeredModules) {
                if (registeredModules[i] && registeredModules[i].dclamd === MODULE_LOADING) {
                    notLoadedModules.push(registeredModules[i]);
                }
            }
            if (notLoadedModules.length) {
                throw new Error("These modules didn't load: " + notLoadedModules.map(function ($) { return $.name; }).join(', '));
            }
        });
    }
    function _toUrl(id, context) {
        var changed = false;
        switch (id) {
            case 'require':
            case 'exports':
            case 'module':
                return id;
        }
        var newContext = (context || settings.baseUrl).split('/');
        newContext.pop();
        var idParts = id.split('/');
        var i = idParts.length;
        while (--i) {
            switch (idParts[0]) {
                case '..':
                    newContext.pop();
                case '.':
                case '':
                    idParts.shift();
                    changed = true;
            }
        }
        return (newContext.length && changed ? newContext.join('/') + '/' : '') + idParts.join('/');
    }
    require.toUrl = _toUrl;
})(loader || (loader = {}));
globalObject.define = loader.define;
globalObject.dclamd = loader;
